Assay config.newformat

pruzanov opened the pull request:
This update includes the following changes:

- Significant re-write of the code to make flask ui work with new assay_info which specifies workflow versions
- Interface is simplified
- applying presets now requires confirmation (Record) as any other change
- flask UI scans olives and append any new deployed versions to configuration
- staged assay configuration produced, needs to be reviewed and edited before deployment
- the code is more robust as it uses the best flask coding practices

For reviewers: please try to update/install flask UI from this branch. I expect this to launch and work in your web browser without any issues. Any suggestions on how to improve the User Experience are welcome.

pruzanov committed:
Pre-release, in the code we still use variables referring project, but we are fully switched to assays

pruzanov committed:
Added check for existing config/version when cloning

pruzanov committed:
Handle special characters (slashes) in URLS properly

pruzanov committed:
made a regex to be compliant with Python 3.12

pruzanov committed:
Ensure we do not generate unicode escape sequences when dumping config to JSON

pruzanov committed:
Re-write of flask UI to accomodate new format of assay_info configuration file

pruzanov committed:
Minor changes

pruzanov committed:
Filter instance-specific workflows when applying presets

avarsava commented on app.py:
```diff
@@ -0,0 +1,332 @@
+    return assay_config
+
+
+""" For state: Load presets TODO: update them here if config changed"""
```
outstanding todo?

pruzanov replied:
That's an older TODO, already implemented

avarsava replied:
OK, should be removed then

avarsava commented on app.py:
```diff
@@ -0,0 +1,332 @@
+state.config = load_config()
+state.preset_list = load_presets()
+
+"""Pretty print json with arrays in single line"""
```
python json stdlib already supports pretty printing using the sort_keys and indent options on json.dumps()

pruzanov replied:
not the way I want it

avarsava commented on app.py:
```diff
@@ -0,0 +1,332 @@
+            jstring = re.sub(r'(\d\")\s+', r'\1', jstring)
+            jstring = re.sub(r'(\.\d\",)\s+', r'\1', jstring)
+            jstring = re.sub(r'(\[)\s+', r'\1', jstring)
+            jstring = re.sub(r'(\d\",)\n', r'\1', jstring)
```
reimplementation of pretty_json()?

pruzanov replied:
It is for printing arrays in one line, I don't think there are any built-in functions which can do this

avarsava replied:
I mean you have reimplemented pretty_json() from line 58 here rather than calling it.

pruzanov replied:
haha that's true, I will fix this

avarsava commented on app.py:
```diff
@@ -0,0 +1,332 @@
+
+""" The App starts here """
+if __name__ == "__main__":
+    my_app.run(debug=True)
```
debug mode left on

avarsava commented on README.md:
```diff
@@ -32,9 +31,8 @@ Normally, you would update your production branch of the repo with shesmu config
- export FLASK_APP=config_ui.py
  export FLASK_ENV=development
- export UICONFIG_SETTINGS="/home/USERNAME/secrets/ui_config.cfg"
+ export UICONFIG_SETTINGS="/home/USERNAME/secrets/ui_config.toml"
```
These instructions conflict with pre_flight.sh

pruzanov replied:
Both README and pre_flight script give an example of a path which needs to be changed, I wouldn't say that is truly a conflict. A user will change this anyways

grapigeau replied:
I also found it confusing that is a path not `CONFIG_FILE_DIRECTORY` as in the pre-flight instructions

avarsava commented on services/olivescan_ops.py:
```diff
@@ -0,0 +1,72 @@
+        vetted_tags = []
+        vetted_names = []
+        try:
+            run_lines = subprocess.check_output(f"grep 'Run ' '{m_olive}'", shell=True).decode().strip()
```
Why shell out to grep (which, while unlikely, may not be installed) instead of using file i/o?

pruzanov replied:
Maybe in a future.... In our environment it works just fine

avarsava commented on services/olivescan_ops.py:
```diff
@@ -0,0 +1,72 @@
+            run_lines = []
+
+        for rl in run_lines:
+            next_tag = re.search(r"v(\d+_\d+_*\d*\w*)$", rl)
+            next_name = re.search(r"(\S+)_v\d+_\d+_*\d*\w*$", rl)
```
next_tag is reimplemented in next_name. If you change to `(\S+)_v(\d+_\d+_*\d*\w*)$` then the name is group 1 and the tag is group 2

avarsava reviewed (commented):
1. It could be easier to manage which changes to save if this staging file went away and changes were made directly to the assay config file. git then provides diffing between the old and new versions natively and the interactive modes of git add can be used to verify the changes.
2. So if an assay is opened and saved without any changes, all the latest versions are appended anyway? Why this design? It seems very prone to adding versions to an assay for which the workflow version is not validated. If you only want to add one workflow and not update versions for everything, then it becomes easier to use a text editor instead of this tool.
3. How is sanity checking implemented? eg if a Workflow B requires the output of Workflow A, and B is enabled but not A, is that flagged anywhere?
4. What do you see as the advantage of parsing olives for workflow versions instead of using the vidarr workflows API?

pruzanov commented:
> 1. It could be easier to manage which changes to save if this staging file went away and changes were made directly to the assay config file. git then provides diffing between the old and new versions natively and the interactive modes of git add can be used to verify the changes.
> 2. So if an assay is opened and saved without any changes, all the latest versions are appended anyway? Why this design? It seems very prone to adding versions to an assay for which the workflow version is not validated. If you only want to add one workflow and not update versions for everything, then it becomes easier to use a text editor instead of this tool.
> 3. How is sanity checking implemented? eg if a Workflow B requires the output of Workflow A, and B is enabled but not A, is that flagged anywhere?
> 4. What do you see as the advantage of parsing olives for workflow versions instead of using the vidarr workflows API?

1. Not sure if it is going to be easier, probably the same as writing into stage config and vimdiff. Another way to do it, yes. Easier - debatable. But writing back to the config is possible with the current setup - we need to point to the assay_info when specifying the staging file in the settings.
2. This is by design and the problem arises only if we write directly back to assay_info as you suggested in 1. Changes are reviewed and approved manually, in most cases we will accept them all b/c otherwise a workflow will stop running.
3. This not addressed by flask UI, a bit advanced to handle by a config-editing tool
4. We have a lot of 'garbage' workflow versions which were deployed in vidarr but never used. When we scan olives we do not have to worry about dealing with those versions

pruzanov committed:
Minor fixes

pruzanov committed:
Changed name for clarity

avarsava commented:
> This is by design and the problem arises only if we write directly back to assay_info as you suggested in 1. Changes are reviewed and approved manually, in most cases we will accept them all b/c otherwise a workflow will stop running.

This will require each person editing the file to have an extensive knowledge of what is and isn't validated in clinical, especially as we start having multiple versions running in parallel to support ramp-down of projects on older assays. I think there is considerable risk of people becoming numb to the updates and approving every change the tool suggests.

pruzanov commented:
> > This is by design and the problem arises only if we write directly back to assay_info as you suggested in 1. Changes are reviewed and approved manually, in most cases we will accept them all b/c otherwise a workflow will stop running.
> 
> This will require each person editing the file to have an extensive knowledge of what is and isn't validated in clinical, especially as we start having multiple versions running in parallel to support ramp-down of projects on older assays. I think there is considerable risk of people becoming numb to the updates and approving every change the tool suggests.

Certainly this mechanism is somewhat high-maintenance, but with proper peer review process and clear guidance on how to perform updates (especially for clinical side) we should be fine.

pruzanov committed:
Monor corrections to README

pruzanov committed:
Update development environment details in README

grapigeau commented on README.md:
```diff
@@ -18,23 +18,21 @@ pip install -r requirements.txt
 
 Modify the new file to point to the locations of the shesmu configuration files (currently in
-spb-seqware-production).
+analysis_config).
```
analysis-config

pruzanov replied:
Thanks, I am mostly interested in hearing if the installation worked for you.

grapigeau replied:
I followed up in response to your slack message but I can add a comment /move discussion here if that's preferable for you

grapigeau commented on ui-config.toml.example:
```diff
@@ -0,0 +1,10 @@
+[data]
+local_olive_dir="$HOME/code/shesmu/"
```
Why include the `/code/` ? That's not part of analysis-config structure

pruzanov replied:
This is just en arbitrary example, I can change it if it is confusing

pruzanov committed:
Fix path to shesmu configuration files in README

Correct the path reference for shesmu configuration files.

pruzanov committed:
Update README with virtual environment instructions

Added instructions for creating a virtual environment with Python 3.12 or newer.

grapigeau commented on README.md:
```diff
@@ -18,23 +20,21 @@ pip install -r requirements.txt
- export FLASK_APP=config_ui.py
  export FLASK_ENV=development
- export UICONFIG_SETTINGS="/home/USERNAME/secrets/ui_config.cfg"
+ export UICONFIG_SETTINGS="/home/USERNAME/secrets/ui_config.toml"
```
should be _ui-config.toml_ to match above

xuemeiluo commented:
1. assay_stage.jsonconfig wasn't found
2. There is an Internal Server Error after selecting presets
3. After I changed a workflow for an assay, the change wasn't made to the assay config

pruzanov commented:
> 1. assay_stage.jsonconfig wasn't found
> 2. There is an Internal Server Error after selecting presets
> 3. After I changed a workflow for an assay, the change wasn't made to the assay config

1: it may not exists when flask UI starts, this is a destination file for writing the updated config
2: I put a copy of presets in new format on slack, but we also can add it to the assay-config
3: that is because changes are written into a file specified by assay_stage_file in configuration .toml file. It is possible to write into assay_info.jsonconfig, but in this case assay_stage_file and assay_config_file should be both set to point to the same file, assay_info.jsonconfig

grapigeau commented:
> > 1. assay_stage.jsonconfig wasn't found
> > 2. There is an Internal Server Error after selecting presets
> > 3. After I changed a workflow for an assay, the change wasn't made to the assay config
> 
> 1: it may not exists when flask UI starts, this is a destination file for writing the updated config 2: I put a copy of presets in new format on slack, but we also can add it to the assay-config 3: that is because changes are written into a file specified by assay_stage_file in configuration .toml file. It is possible to write into assay_info.jsonconfig, but in this case assay_stage_file and assay_config_file should be both set to point to the same file, assay_info.jsonconfig

For point 3, can this be documented somewhere? Otherwise looks good to me.

pruzanov committed:
Added presets validation code

pruzanov committed:
Minor cosmetic change

pruzanov committed:
Added Overview functionality

pruzanov committed:
Added ability to overview configuration in two modes

pruzanov committed:
Updating documentation

pruzanov committed:
Support multiple prefixes when identifying instance-specific assays

pruzanov committed:
Added manual file

pruzanov committed:
Allow multiple prefixes to identify instance-specific assays

pruzanov committed:
Insert available wf version into configs of all assay versions, not only the latest

pruzanov committed:
Added Gencode support

pruzanov committed:
Restructured assy_info support
